File: app/javascript/application.js
/* Copyright 2023-2024, Pablo Fernandez
 *
 * This file is part of Repeater World.
 *
 * Repeater World is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General
 * Public License as published by the Free Software Foundation, either version 3 of the License.
 *
 * Repeater World is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied 
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more 
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along with Repeater World. If not, see
 * <https://www.gnu.org/licenses/>.
 */

// Configure your import map in config/importmap.rb. Read more: https://github.com/rails/importmap-rails

// Set up Sentry as early as possible, so any errors are caught.
import * as Sentry from "@sentry/browser"
import {BrowserTracing} from "@sentry/tracing"
// Make FontAwesome 6 work... but why?
// https://stackoverflow.com/questions/71430573/can-font-awesome-be-used-with-importmaps-in-rails-7
// https://discuss.rubyonrails.org/t/can-font-awesome-be-used-with-importmaps-in-rails-7/80238
import {far} from "@fortawesome/free-regular-svg-icons"
import {fas} from "@fortawesome/free-solid-svg-icons"
import {fab} from "@fortawesome/free-brands-svg-icons"
import {library} from "@fortawesome/fontawesome-svg-core"
import "@fortawesome/fontawesome-free"
// Other imports
import "@hotwired/turbo-rails"

// And now, our code.
import "controllers"

if (SENTRY_DSN !== null) {
  let environment = RAILS_ENV
  if (RENDER === "true") {
    environment = IS_PULL_REQUEST === "true" ? "review" : RAILS_ENV
  }

  Sentry.init({
    dsn: SENTRY_DSN,
    integrations: [new BrowserTracing()],
    tracesSampleRate: 0.1, // Reduce this as the application has more traffic.
    environment: environment
  })

  if (IS_PULL_REQUEST === "true") {
    Sentry.configureScope(function (scope) {
      if (RENDER_SERVICE_NAME) {
        let pr = RENDER_SERVICE_NAME.split("-")
        pr = pr[pr.length - 1]
        scope.setTag("pr", pr)
      }
    })
  }
} else {
  if (RAILS_ENV === "production") {
    console.log("Missing SENTRY_DSN")
  }
}
// End of sentry setup.

library.add(far, fas, fab)
// End of FontAwesome setup

File: config/application.rb
# Copyright 2023-2024, Pablo Fernandez
#
# This file is part of Repeater World.
#
# Repeater World is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General
# Public License as published by the Free Software Foundation, either version 3 of the License.
#
# Repeater World is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
# warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along with Repeater World. If not, see
# <https://www.gnu.org/licenses/>.

require_relative "boot"

require "rails"
# Pick the frameworks you want:
require "active_model/railtie"
require "active_job/railtie"
require "active_record/railtie"
require "active_storage/engine"
require "action_controller/railtie"
require "action_mailer/railtie"
require "action_mailbox/engine"
require "action_text/engine"
require "action_view/railtie"
require "action_cable/engine"
# require "rails/test_unit/railtie"

# Require the gems listed in Gemfile, including any gems
# you've limited to :test, :development, or :production.
Bundler.require(*Rails.groups)

module RepeaterWorld
  class Application < Rails::Application
    # Initialize configuration defaults for originally generated Rails version.
    config.load_defaults 7.0

    # Configuration for the application, engines, and railties goes here.
    #
    # These settings can be overridden in specific environments using the files
    # in config/environments, which are processed later.
    #
    # config.time_zone = "Central Time (US & Canada)"
    # config.eager_load_paths << Rails.root.join("extras")

    # https://stackoverflow.com/questions/17951037/how-to-require-some-lib-files-from-anywhere
    config.autoload_paths += %W[#{config.root}/lib]
    config.autoload_paths += Dir["#{config.root}/lib/**/"]

    # Don't generate system test files.
    config.generators.system_tests = nil

    # https://guides.rubyonrails.org/action_mailer_basics.html#previewing-emails
    config.action_mailer.preview_paths << "#{Rails.root}/spec/mailers/previews/"

    # Necessary for handling 404 errors, so that /404 gets called as a catch-all
    config.exceptions_app = routes

    # https://lab.io/articles/2017/04/13/uuids-rails-5-1/
    config.generators do |g|
      g.orm :active_record, primary_key_type: :uuid
    end
  end
end
